bioconda_utils/lint_functions.py

```diff
@@ -0,0 +1,207 @@
+tuple so that other code knows it's a QC check function.
+
+(Note: even though we can parse `meta` from `recipe`, both are passed to speed
+things up. We need the recipe dir to do things like check for test scripts, and
+we only want to parse the meta.yaml once and pass it to check functions.  The
+dataframe takes a little time to download, and some checks need it, so we just
+do it once and pass it around)
+"""
+
+import os
+import glob
+
+
+def _subset_df(recipe, meta, df):
+    """
+    Helper function to get the subset of `df` for this recipe.
+    """
+    name = meta['package']['name']
+    version = meta['package']['version']
+
+    # Some recipes have an empty build section, so it'll be None and we can't
+    # do a chained get.
+    build_number = 0
+    build_section = meta.get('build')
+    if build_section:
+        build_number = int(build_section.get('number', 0))
+    return df[
+        (df.name == name)
+        & (df.version == version)
+        & (df.build_number == build_number)
+    ]
+
+
+def in_other_channels(recipe, meta, df):
+    """
+    Does the package exist in any other non-bioconda channels?
+    """
+    results = _subset_df(recipe, meta, df)
+    channels = set(results.channel).difference(['bioconda'])
+    if len(channels):
+        return {
+            'exists_in_channels': channels,
+            'fix': 'consider deprecating',
+        }
+
+
+def already_in_bioconda(recipe, meta, df):
+    """
+    Does the package exist in bioconda?
+    """
+    results = _subset_df(recipe, meta, df)
+    channels = set(results.channel)
+    if 'bioconda' in channels:
+        return {
+            'already_in_bioconda': True,
+            'fix': 'bump version or build number'
+        }
+
+
+def missing_home(recipe, meta, df):
+    try:
+        meta['about']['home']
+    except KeyError:
+        return {
+            'missing_home': True,
+            'fix': 'add about:home',
+        }
+
+
+def missing_summary(recipe, meta, df):
+    try:
+        meta['about']['summary']
+    except KeyError:
+        return {
+            'missing_summary': True,
+            'fix': 'add about:summary',
+        }
+
+
+def missing_license(recipe, meta, df):
+    try:
+        meta['about']['license']
+    except KeyError:
+        return {
+            'missing_license': True,
+            'fix': 'add about:license'
+        }
+
+
+def missing_tests(recipe, meta, df):
+    test_files = ['run_test.py', 'run_test.sh', 'run_test.pl']
+    if 'test' not in meta:
+        if not any([os.path.exists(os.path.join(recipe, f)) for f in test_files]):
+            return {
+                'no_tests': True,
+                'fix': 'add basic tests',
+            }
+
+
+def missing_hash(recipe, meta, df):
+    # could be a meta-package if no source section or if None
+    try:
+        src = meta['source']
+        if src is None:
+            return
+    except KeyError:
+        return
+
+
+    if not any(
+        (
+            'md5' in src,
```

Comment:
Sometimes, md5 is even stronger than sha256 in my opinion. This is the case if md5 is provided by upstream. I think it is always better to use the upstream hash than a hash that you have to compute on your own, because at that time you already might have downloaded a corrupted source.